cartographer_rviz/cartographer_rviz/drawable_submap.cc

```diff
@@ -159,13 +168,16 @@ void DrawableSubmap::SetAlpha(const double current_tracking_z) {
 
 void DrawableSubmap::UpdateSceneNode() {
   ::cartographer::common::MutexLocker locker(&mutex_);
+  render_in_progress_ = false;
+  if (quitting_) {
```

Comment:
Hi! Sorry that I was not specific enough. You are exactly right about everything. I agree, your solution is more elegant; by using the `wait` method of the request future object, there is no need to split the `query_in_progress_` flag. However, it still might be more accurate to rename the existing flag (without splitting it) to "query or render in progress" (or something more specific instead of "render", such as "texture update"), because effectively it really behaves that way.

> You have to be more specific for me to follow your reasoning. For example, you set quitting_ to true in the destructor, but use it in MaybeFetchTexture() line 125. Why would this member function be called if the object was destroyed? It seems this should and does never happen. Line 172 seems to defend against the fact that the object to which this points was already destroyed. At this point looking at quitting_ is too late as the object is already gone.

That is right. Adding the check for `quitting_` was a precautionary (more like desperate :-) ) measure in case we would be extremely unlucky and the destructor is invoked just after fetch completes and emits the `RequestSucceeded()` signal, which would be caught by Qt. The destructor and `UpdateSceneNode()` both compete to lock the mutex; destructor wins and deletes the object. If `UpdateSceneNode()` now reads from a deleted object, the least I could do is read the quitting boolean, which should still be intact, and we know that we should give up. If you're certain that this scenario is impossible, great! We can then remove `quitting_` altogether.

To summarize:
1) Remove `render_in_progress_`, use `rpc_request_future_.wait()`, keep existing `query_in_progress_` as it is, possibly rename it to reference that it also includes updating the textures
2) In `SubmapsDisplay::update()`, I stil think that we should have the scope of the locker for the whole function, because we are also iterating through `trajectories_` in the second part (you weren't suggesting to remove that, only the second commit, correct?)
3) If you're certain that the signal emitted before/during `DrawableSubmap` destruction by the fetcher won't result in a call of `UpdateSceneNode()` where it acquires the mutex after destruction, we can remove `quitting_`.